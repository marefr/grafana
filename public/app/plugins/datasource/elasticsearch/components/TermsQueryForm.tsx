import React, { PureComponent, ChangeEvent } from 'react';
import { Select, SelectOptionItem, FormLabel, Input } from '@grafana/ui';

export interface TermsQueryFormProps {
  query: any;
  fields: SelectOptionItem[];
  onChange: (query: any, definition: string) => void;
}

export interface TermsQueryFormState {
  field: string;
  query: string;
  size: string;
  orderBy: string;
  order: string;
}

const defaultState: TermsQueryFormState = {
  field: '',
  query: '',
  size: '',
  orderBy: '_term',
  order: 'asc',
};

const orderByOptions: SelectOptionItem[] = [
  { value: '_term', label: 'Term value' },
  { value: '_count', label: 'Doc count' },
];

const orderOptions: SelectOptionItem[] = [{ value: 'asc', label: 'Ascending' }, { value: 'desc', label: 'Descending' }];

export class TermsQueryForm extends PureComponent<TermsQueryFormProps, TermsQueryFormState> {
  constructor(props: TermsQueryFormProps) {
    super(props);
    this.state = {
      ...defaultState,
      ...props.query,
      size: props.query.size ? props.query.size.toString() : '',
    };
  }

  triggerChange() {
    const { onChange } = this.props;
    const { field, query, size, orderBy, order } = this.state;
    const termsQuery: any = {
      find: 'terms',
      field,
    };

    if (defaultState.field === field) {
      return;
    }

    if (defaultState.query !== query) {
      termsQuery.query = query;
    }

    if (defaultState.size !== size) {
      termsQuery.size = Number.parseInt(size, 10);
    }

    if (defaultState.orderBy !== orderBy) {
      termsQuery.orderBy = orderBy;
    }

    if (defaultState.order !== order) {
      termsQuery.order = order;
    }

    onChange(termsQuery, `Terms(${field})`);
  }

  onFieldChange = (item: SelectOptionItem) => {
    this.setState(
      {
        field: item.value,
      },
      () => {
        this.triggerChange();
      }
    );
  };

  onQueryChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({
      query: event.target.value,
    });
  };

  onQueryBlur = () => {
    this.triggerChange();
  };

  onSizeChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({
      size: event.target.value,
    });
  };

  onSizeBlur = () => {
    this.triggerChange();
  };

  onOrderByChange = (item: SelectOptionItem) => {
    let { order } = this.state;
    const orderBy = item.value;

    if (orderBy === '_count') {
      order = 'desc';
    }

    this.setState(
      {
        orderBy,
        order,
      },
      () => {
        this.triggerChange();
      }
    );
  };

  onOrderChange = (item: SelectOptionItem) => {
    this.setState(
      {
        order: item.value,
      },
      () => {
        this.triggerChange();
      }
    );
  };

  render() {
    const { fields } = this.props;
    const { field, query = '', size, orderBy, order } = this.state;
    return (
      <>
        <div className="form-field">
          <FormLabel className="query-keyword">Field</FormLabel>
          <Select
            placeholder="Choose field"
            options={fields}
            value={fields.find(o => o.value === field)}
            onChange={this.onFieldChange}
            width={11}
          />
        </div>
        <div className="form-field">
          <FormLabel className="query-keyword">Query</FormLabel>
          <Input placeholder="lucene query" onBlur={this.onQueryBlur} onChange={this.onQueryChange} value={query} />
        </div>
        <div className="form-field">
          <FormLabel className="query-keyword">Size</FormLabel>
          <Input
            className="gf-form-input width-6"
            placeholder="500"
            onBlur={this.onSizeBlur}
            onChange={this.onSizeChange}
            type="number"
            value={size}
          />
        </div>
        <div className="form-field">
          <FormLabel className="query-keyword">Order By</FormLabel>
          <Select
            placeholder="Choose order"
            options={orderByOptions}
            value={orderByOptions.find(o => o.value === orderBy)}
            onChange={this.onOrderByChange}
            width={11}
          />
          <Select
            placeholder="Choose order"
            options={orderOptions}
            value={orderOptions.find(o => o.value === order)}
            onChange={this.onOrderChange}
            width={11}
          />
        </div>
      </>
    );
  }
}
