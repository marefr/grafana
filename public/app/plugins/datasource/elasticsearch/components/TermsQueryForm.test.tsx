import React from 'react';
import { shallow } from 'enzyme';
import { TermsQueryForm } from './TermsQueryForm';

const setup = (propOverrides?: object) => {
  const props = {
    onChange: () => {},
    query: {},
    fields: [{ value: 'field.a', label: 'field.a', description: 'number' }],
    ...propOverrides,
  };

  const wrapper = shallow(<TermsQueryForm {...props} />);
  const instance = wrapper.instance() as TermsQueryForm;

  return {
    wrapper,
    instance,
  };
};

describe('TermsQueryForm', () => {
  it('should render component', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('when loading component should not trigger change event', () => {
    const props = {
      onChange: jest.fn(),
    };
    setup(props);
    expect(props.onChange.mock.calls.length).toBe(0);
  });

  it('when changing field should trigger change event', () => {
    const props = {
      onChange: jest.fn(),
    };
    const { instance } = setup(props);
    instance.onFieldChange({ value: 'field.a', label: 'field.a' });
    expect(props.onChange.mock.calls.length).toBe(1);
    expect(props.onChange.mock.calls[0][0].find).toBe('terms');
    expect(props.onChange.mock.calls[0][0].field).toBe('field.a');
    expect(props.onChange.mock.calls[0][1]).toBe('Terms(field.a)');
  });

  it('when changing query should trigger change event', () => {
    const props = {
      onChange: jest.fn(),
      query: { field: 'field.a' },
    };
    const { instance } = setup(props);
    instance.onQueryChange({
      target: {
        value: 'field.a:*',
      },
    } as any);
    instance.onQueryBlur();
    expect(props.onChange.mock.calls.length).toBe(1);
    expect(props.onChange.mock.calls[0][0]).toEqual({
      find: 'terms',
      field: 'field.a',
      query: 'field.a:*',
    });
    expect(props.onChange.mock.calls[0][1]).toBe('Terms(field.a)');
  });

  it('when changing size should trigger change event', () => {
    const props = {
      onChange: jest.fn(),
      query: { field: 'field.a' },
    };
    const { instance } = setup(props);
    instance.onSizeChange({
      target: {
        value: '1000',
      },
    } as any);
    instance.onSizeBlur();
    expect(props.onChange.mock.calls.length).toBe(1);
    expect(props.onChange.mock.calls[0][0]).toEqual({
      find: 'terms',
      field: 'field.a',
      size: 1000,
    });
    expect(props.onChange.mock.calls[0][1]).toBe('Terms(field.a)');
  });

  it('when changing order by to term value should trigger change event', () => {
    const props = {
      onChange: jest.fn(),
      query: { field: 'field.a' },
    };
    const { instance } = setup(props);
    props.onChange.mockReset();
    instance.onOrderByChange({ value: '_term', label: 'Term value' });
    expect(props.onChange.mock.calls.length).toBe(1);
    expect(props.onChange.mock.calls[0][0]).toEqual({
      find: 'terms',
      field: 'field.a',
    });
    expect(props.onChange.mock.calls[0][1]).toBe('Terms(field.a)');
  });

  it('when changing order by to doc count should trigger change event', () => {
    const props = {
      onChange: jest.fn(),
      query: { field: 'field.a' },
    };
    const { instance } = setup(props);
    props.onChange.mockReset();
    instance.onOrderByChange({ value: '_count', label: 'Doc Count' });
    expect(props.onChange.mock.calls.length).toBe(1);
    expect(props.onChange.mock.calls[0][0]).toEqual({
      find: 'terms',
      field: 'field.a',
      orderBy: '_count',
      order: 'desc',
    });
    expect(props.onChange.mock.calls[0][1]).toBe('Terms(field.a)');
  });

  it('when changing order to descending should trigger change event', () => {
    const props = {
      onChange: jest.fn(),
      query: { field: 'field.a' },
    };
    const { instance } = setup(props);
    props.onChange.mockReset();
    instance.onOrderChange({ value: 'desc', label: 'Descending' });
    expect(props.onChange.mock.calls.length).toBe(1);
    expect(props.onChange.mock.calls[0][0]).toEqual({
      find: 'terms',
      field: 'field.a',
      order: 'desc',
    });
    expect(props.onChange.mock.calls[0][1]).toBe('Terms(field.a)');
  });
});
