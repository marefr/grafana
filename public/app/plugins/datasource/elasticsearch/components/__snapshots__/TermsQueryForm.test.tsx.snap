// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TermsQueryForm should render component 1`] = `
<Fragment>
  <div
    className="form-field"
  >
    <Component
      className="query-keyword"
    >
      Field
    </Component>
    <Select
      autoFocus={false}
      backspaceRemovesValue={true}
      className=""
      components={
        Object {
          "Group": [Function],
          "IndicatorsContainer": [Function],
          "MenuList": [Function],
          "Option": [Function],
          "SingleValue": [Function],
        }
      }
      isClearable={false}
      isDisabled={false}
      isLoading={false}
      isMulti={false}
      isSearchable={true}
      maxMenuHeight={300}
      menuIsOpen={false}
      onChange={[Function]}
      openMenuOnFocus={false}
      options={
        Array [
          Object {
            "description": "number",
            "label": "field.a",
            "value": "field.a",
          },
        ]
      }
      placeholder="Choose field"
      width={11}
    />
  </div>
  <div
    className="form-field"
  >
    <Component
      className="query-keyword"
    >
      Query
    </Component>
    <Input
      className=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="lucene query"
      value=""
    />
  </div>
  <div
    className="form-field"
  >
    <Component
      className="query-keyword"
    >
      Size
    </Component>
    <Input
      className="gf-form-input width-6"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="500"
      type="number"
      value=""
    />
  </div>
  <div
    className="form-field"
  >
    <Component
      className="query-keyword"
    >
      Order By
    </Component>
    <Select
      autoFocus={false}
      backspaceRemovesValue={true}
      className=""
      components={
        Object {
          "Group": [Function],
          "IndicatorsContainer": [Function],
          "MenuList": [Function],
          "Option": [Function],
          "SingleValue": [Function],
        }
      }
      isClearable={false}
      isDisabled={false}
      isLoading={false}
      isMulti={false}
      isSearchable={true}
      maxMenuHeight={300}
      menuIsOpen={false}
      onChange={[Function]}
      openMenuOnFocus={false}
      options={
        Array [
          Object {
            "label": "Term value",
            "value": "_term",
          },
          Object {
            "label": "Doc count",
            "value": "_count",
          },
        ]
      }
      placeholder="Choose order"
      value={
        Object {
          "label": "Term value",
          "value": "_term",
        }
      }
      width={11}
    />
    <Select
      autoFocus={false}
      backspaceRemovesValue={true}
      className=""
      components={
        Object {
          "Group": [Function],
          "IndicatorsContainer": [Function],
          "MenuList": [Function],
          "Option": [Function],
          "SingleValue": [Function],
        }
      }
      isClearable={false}
      isDisabled={false}
      isLoading={false}
      isMulti={false}
      isSearchable={true}
      maxMenuHeight={300}
      menuIsOpen={false}
      onChange={[Function]}
      openMenuOnFocus={false}
      options={
        Array [
          Object {
            "label": "Ascending",
            "value": "asc",
          },
          Object {
            "label": "Descending",
            "value": "desc",
          },
        ]
      }
      placeholder="Choose order"
      value={
        Object {
          "label": "Ascending",
          "value": "asc",
        }
      }
      width={11}
    />
  </div>
</Fragment>
`;
